# Homework 1 (100 pts)
## Applied AI/Machine Learning (ITSS 4382)

Deliverables:

- Submit your answers to conceptual questions (described in another .pdf file) in a pdf file
- Write down your codes in the given cells described in this file, denoted as "YOUR CODE HERE"
- Write down your discussion in the given cells, denoted as "YOUR DISCUSSION HERE"
- Feel free to add any blocks if needed using ALT+ENTER
- <b style="color: red">Submit both .pdf and .ipynb files to eLearning</b>


This assignment covers Supervised Learning models. In this assignment, you are required to train both classification and Regression models.

In [1]:
NAME = "Winston Shih"

In [2]:
%matplotlib inline 
import numpy as np
import pandas as pd
# packages for plotting
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

# Part I. KNN Classification 

In this part, you are required to use the [Breast Cancer Wisconsin (Diagnostic) Database](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_breast_cancer.html#sklearn.datasets.load_breast_cancer) to create a classifier that can help diagnose patients. This dataset is a classic and very easy binary classification dataset. Read more in the [User Guide](https://scikit-learn.org/stable/datasets/toy_dataset.html#breast-cancer-dataset).

In [3]:
# Generating Data
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()
print(cancer.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


## Q1a | What features does the data have? (3 pts) 

In [4]:
# List all the features in the data file
# WRITE YOUR CODE HERE
print(cancer.feature_names)


['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


## Q1b | What values does the target have? (3 pts) 

In [5]:
# List all the values in the target
# WRITE YOUR CODE HERE
cancer.target


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,

## Q1c | Show the number records in the dataset. (3 pts) 

In [6]:
# Count all the rows only in the data file
# WRITE YOUR CODE HERE 
cancer.data.shape[0]


569

## Q1d | Split the data with 75% training, 25% test, and random_state=42. (3 pts) 

In [7]:
# Splitting Data
# WRITE YOUR CODE HERE
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(cancer.data, cancer.target, random_state=42)



## Classification Model 

## Q2 (20 points)
- Train a k-NN model with grid search and cross-validation. Let the choices of k be: [5, 10, 15, 20, 25]. Use 5-fold cross-validation.
- Report (1) the best k chosen, (2) the test accuracy under the best model, and (3) the mean validation accuracy through the cross-validation process (under the best model). *Hint: the mean validation accuracy can be obtained through MODEL_NAME.best_score_*

In [9]:
# WRITE YOUR CODE HERE
from sklearn.model_selection import GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
param_knn = {'n_neighbors': [5, 10, 15, 20, 25]}
knn = KNeighborsClassifier()
gridsearch_knn = GridSearchCV(knn, param_knn, cv=5)
gridsearch_knn.fit(X_train, y_train)











GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [5, 10, 15, 20, 25]})

In [10]:
# WRITE YOUR CODE HERE
gridsearch_knn.best_params_, gridsearch_knn.score(X_train, y_train), gridsearch_knn.best_score_,



({'n_neighbors': 10}, 0.9295774647887324, 0.9270861833105336)

<br><br><br><br>
# Part II. Regression Models

In this part, you are required to analyze the daily version of the [Capital Bikeshare System dataset](https://archive.ics.uci.edu/ml/datasets/Bike+Sharing+Dataset) from the UCI Machine Learning Repository. This data set contains information about the daily count of bike rental checkouts in Washington, D.C.’s bikeshare program between 2011 and 2012. It also includes information about the weather and seasonal/temporal features for that day (e.g., if it was a weekday).
- **day:** Day of the record (relative to day 1: 2011-01-01)
- **season:** Season (1: spring, 2: summer, 3: fall, 4: winter)
- **weekday:** Day of the week (0 = Sunday, 6 = Saturday)
- **workingday:** If day is neither weekend nor holiday is 1, otherwise is 0.
- **weathersit:**<br>
    - 1: Clear, Few clouds, Partly cloudy, Partly cloudy
    - 2: Mist + Cloudy, Mist + Broken clouds, Mist + Few clouds, Mist
    - 3: Light Snow, Light Rain + Thunderstorm + Scattered clouds, Light Rain + Scattered
clouds
- **temp:** Normalized temperature in Celsius
- **windspeed:** Normalized wind speed
- **casual:** Count of checkouts by casual/non-registered users
- **registered:** Count of checkouts by registered users
- **cnt:** Total checkouts

## Preliminaries

Run the cells below to load the split the data. We will use features ('weekday', 'season','workingday', 'temp', 'windspeed', 'weathersit') to predict the checkout counts. We are interested in predicting the total checkout counts (i.e., 'cnt').

In [11]:
day = pd.read_csv("day.csv")
print(day.columns.values) # to find variable names

['instant' 'day' 'season' 'weekday' 'workingday' 'weathersit' 'temp'
 'windspeed' 'casual' 'registered' 'cnt']


In [12]:
day.head()

instant       day  season  weekday  workingday  weathersit      temp  \
0        1  1/1/2011       1        6           0           2  0.344167   
1        2  1/2/2011       1        0           0           2  0.363478   
2        3  1/3/2011       1        1           1           1  0.196364   
3        4  1/4/2011       1        2           1           1  0.200000   
4        5  1/5/2011       1        3           1           1  0.226957   

   windspeed  casual  registered   cnt  
0   0.160446     331         654   985  
1   0.248539     131         670   801  
2   0.248309     120        1229  1349  
3   0.160296     108        1454  1562  
4   0.186900      82        1518  1600

In [13]:
var = ['weekday', 'season','workingday', 'temp', 'windspeed', 'weathersit']
X = day[var]
y = day['cnt']

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

## Q3a | Lasso Regression (20 pts)

- Train a Lasso regression model using the training set with alpha = 50. We would like to use MSE as the performance measure. What is the MSE for the test set?  What is the MSE for training set?

In [14]:
# WRITE YOUR CODE HERE
from sklearn.linear_model import Lasso
lasso_model = Lasso(alpha = 50)
lasso_model.fit(X_train, y_train)
y_pred_test = lasso_model.predict(X_test)
error_lasso = y_test - y_pred_test
MSE_test = np.sum(error_lasso**2)/y_test.shape[0]
print("Mean Squared Error for test set :", MSE_test)
y_pred_train = lasso_model.predict(X_train)
error_train= y_train - y_pred_train
MSE_train = np.sum(error_train**2)/y_train.shape[0]
print("Mean Squared Error for train set :", MSE_train)


















Mean Squared Error for test set : 2042360.3510422485
Mean Squared Error for train set : 1960999.2636079749


## Q3b | Parameters (5 pts)
- Report the coefficients and the intercept of the above Lasso regression.

In [15]:
# WRITE YOUR CODE HERE
Intercept_lasso =  lasso_model.intercept_ 
coefficient_lasso = lasso_model.coef_  
print("The intercept is:", Intercept_lasso)
print("The coefficients are:", coefficient_lasso)












The intercept is: 2242.2069585174436
The coefficients are: [  66.42091432  484.36577205    0.         3785.62747889   -0.
 -698.48336419]


## Q3c | Report (10 pts)

- Based on the current model, which feature(s) are not useful in predicting the outcome (y)? Explain Briefly. **Maximum 3 sentences**

<br><br>
#### 'workingday' and 'windspeed' are not useful in predicting outcome (y) because they have a zero coefficient. A zero coefficient means the features don't affect the outcome variable. Meanwhile, features with non zero coefficients like 'weekday' and 'season' are useful at predicting outcome (y).
<br><br><br><br><br><br><br>

## Q4 | Ridge Regression (10 pts)

- For the same dataset, train a Ridge regression model with lambda (i.e., hyperparameter) = 0.05. Report the coefficient and intercept of the Ridge model.

In [16]:
# WRITE YOUR CODE HERE
from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.05)
ridge_reg.fit(X_train, y_train)
print("The intercept is:", ridge_reg.intercept_)
print("The coefficients are:", ridge_reg.coef_)









The intercept is: 2093.039774263602
The coefficients are: [   76.7782694    413.94962299   132.66594864  5230.10542569
 -1966.83827101  -801.27644575]


<br><br>After inspecting the data, we have found that the total checkouts is the sum of the casual checkouts and registered checkouts. That is, cnt = casual + registered. We believe that the checkout behavior of the two groups of users (i.e., registered vs. casual) would be significantly different. Therefore, we want to predict casual and registered checkout separately. For simplicity, we use linear regression models.

Run the cell below for data splitting. *(Hint: prefix `yr_` represents y for registered, prefix `yc_` represents y for casual, prefix `yt_` represents y for total checkout).*

In [17]:
var = ['weekday', 'season','workingday', 'temp', 'windspeed', 'weathersit']
var_ys = ['registered', 'casual', 'cnt']

X = day[var]
ys = day[var_ys]

from sklearn.model_selection import train_test_split
X_train, X_test, ys_train, ys_test = train_test_split(X, ys, test_size=0.25, random_state=42)

yr_train, yc_train, yt_train = ys_train['registered'], ys_train['casual'], ys_train['cnt']
yr_test, yc_test, yt_test  = ys_test['registered'], ys_test['casual'], ys_test['cnt']

## Q5a | Linear Regression (10 pts)

- Train a linear model. Let X be ['weekday', 'season','workingday', 'temp', 'windspeed', 'weathersit']. Let y be registered checkout count. Report the coefficients of variables (no need to report the intercept).

In [18]:
# Registered checkouts
# WRITE YOUR CODE HERE
from sklearn.linear_model import LinearRegression
lr_model_registered = LinearRegression()
lr_model_registered.fit(X_train, yr_train)
print("The coefficients for Registered checkouts are:", lr_model_registered.coef_)











The coefficients for Registered checkouts are: [   42.46792561   386.41234982   964.88084644  3240.87682355
 -1290.66064042  -590.2540968 ]


## Q5b | Linear Regression Extended (5 pts)
- Train another linear model. Let X be ['weekday', 'season','workingday', 'temp', 'windspeed', 'weathersit']. Let y be casual checkout count. Report the coefficients of variables (no need to report the intercept).

In [19]:
# WRITE YOUR CODE HERE
lr_model_casual = LinearRegression()
lr_model_casual.fit(X_train, yc_train)
print("The coefficients for casual checkout are:",lr_model_casual.coef_)







The coefficients for casual checkout are: [  34.30634028   26.25046494 -832.67101663 2004.60884834 -705.0750121
 -210.35534097]


## Q5c | Explanation 8 pts)
- Based on the results, we find that one variable has opposite impact on casual checkout count and registered checkout count. What is the variable? Explain briefly. **Maximum 2 sentences**
<br><br><br>

#### 'workingday' is the variable that has an opposite impact on casual checkout count and registered checkout count. For each unit increase in 'workingday' casual checkout count decreases by 832.67 units and registered checkout count decreases 964.88 units.


<br><br><br><br>
